File: MyLeetCode/python/Linked List Random Node.py
###
# Linked List
# https://leetcode.com/problems/linked-list-random-node/discuss/85659/Brief-explanation-for-Reservoir-Sampling
# Time Complexity: O(n)
# Space Complexity:O(n)
###
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):

    def __init__(self, head):
        """
        @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
        :type head: ListNode
        """
        self.head = head
        

    def getRandom(self):
        """
        Returns a random node's value.
        :rtype: int
        """
        num = 1
        cur = self.head.next
        re = self.head
        while cur:
            r = random.randint(0, num)
            if r == 0:
                re = cur
            num += 1
            cur = cur.next
        return re.val


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()
                                        
                            
                    

File: MyLeetCode/Leetcode/Linked List Random Node.cpp
/*Reservoir Sampling
https://discuss.leetcode.com/topic/53753/brief-explanation-for-reservoir-sampling/2
Time Complexity: O(n)
Space Complexity: O(1)
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /** @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node. */
    Solution(ListNode* head) {
        list = head;
    }
    
    /** Returns a random node's value. */
    int getRandom() {
        int num = 2;
        int re = list->val;
        ListNode* node = list->next;
        while (node) {
            int randnum = rand() % num;
            if (randnum == 0) {
                re = node->val;
            }
            num++;
            node = node->next;
        }
        return re;
    }
private:
    ListNode* list;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */